daml-script/dump/src/main/scala/com/daml/script/dump/TreeUtils.scala

```diff
@@ -159,39 +160,97 @@ object TreeUtils {
   def createdReferencedCids(ev: CreatedEvent): Set[ContractId] =
     ev.createArguments.foldMap(args => args.fields.foldMap(f => valueCids(f.getValue.sum)))
 
-  /** A simple event causes the creation of a single contract and returns its contract id.
+  def cmdReferencedCids(cmd: Command): Set[ContractId] = {
+    cmd match {
+      case CreateCommand(createdEvent) =>
+        createdReferencedCids(createdEvent)
+      case ExerciseCommand(exercisedEvent) =>
+        exercisedEvent.choiceArgument.foldMap(arg => valueCids(arg.sum)) + ContractId(
+          exercisedEvent.contractId
+        )
+      case CreateAndExerciseCommand(createdEvent, exercisedEvent) =>
+        createdReferencedCids(createdEvent) ++ exercisedEvent.choiceArgument.foldMap(arg =>
+          valueCids(arg.sum)
+        )
+    }
+  }
+
+  sealed trait Command
+  final case class CreateCommand(createdEvent: CreatedEvent) extends Command
+  final case class ExerciseCommand(exercisedEvent: ExercisedEvent) extends Command
+  final case class CreateAndExerciseCommand(
+      createdEvent: CreatedEvent,
+      exercisedEvent: ExercisedEvent,
+  ) extends Command
+
+  object Command {
+    def fromTree(tree: TransactionTree): Seq[Command] = {
+      val commands = mutable.LinkedHashMap.empty[EventId, Command]
+      val created = mutable.HashMap.empty[ContractId, EventId]
+      tree.rootEventIds.foreach { eventId =>
+        val event = tree.eventsById(eventId)
+        event.kind match {
+          case Kind.Empty =>
+          case Kind.Created(createdEvent) =>
+            commands += EventId(eventId) -> CreateCommand(createdEvent)
+            created += ContractId(createdEvent.contractId) -> EventId(eventId)
+          case Kind.Exercised(exercisedEvent) =>
+            created.get(ContractId(exercisedEvent.contractId)) match {
```

Comment:
https://github.com/digital-asset/daml/issues/9155

Reply:
Good point, the logic only applies to the root events of the tree. I was only taking `create`, `createAndExercise`, and `exercise` into consideration. In which case, I don't think this would pose an issue. However, I forgot about `exerciseByKeyCmd`. Using that it is possible to construct root events where the ordering matters as described here. See https://github.com/digital-asset/daml/issues/9155#issuecomment-800924621.

Reply:
The roots that you can use when submitting are not identical to the roots that you will get on a transaction tree. The projection my party sees can have nodes as roots which were not roots originally.

Reply:
I've changed it so that only adjacent create & exercise events can be converted to `createAndExerciseCmd` and added a test-case for this situation.